File: back-end/ShopHangTet/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using ShopHangTet.Data;
using ShopHangTet.Models;
using ShopHangTet.DTOs;
using ShopHangTet.Services;
using Google.Apis.Auth;
using Microsoft.AspNetCore.Authorization;
using System.Security.Claims;

namespace ShopHangTet.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class AuthController : ControllerBase
    {
        private readonly ShopHangTetDbContext _context;
        private readonly IJwtService _jwtService;
        private readonly IConfiguration _configuration; //IConfig to let  Google Login reads ClientId
        private readonly IEmailService _emailService;

        public AuthController(ShopHangTetDbContext context, IJwtService jwtService, IConfiguration configuration, IEmailService emailService)
        {
            _context = context;
            _jwtService = jwtService;
            _configuration = configuration;
            _emailService = emailService;
        }

        // ===========================
        // 1. REGISTER
        // ===========================
        [HttpPost("register")]
        public async Task<IActionResult> Register([FromBody] RegisterDto request)
        {
            if (await _context.Users.AnyAsync(u => u.Email == request.Email))
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Email này đã được sử dụng."));
            }

            string otpCode = new Random().Next(100000, 999999).ToString();

            var newUser = new UserModel
            {
                Email = request.Email,
                FullName = request.FullName,
                Phone = request.Phone,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(request.Password),
                Role = UserRole.MEMBER,
                Status = UserStatus.ACTIVE,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow,
                IsEmailVerified = false,
                OtpCode = otpCode,
                OtpExpiry = DateTime.UtcNow.AddMinutes(5)
            };

            _context.Users.Add(newUser);
            await _context.SaveChangesAsync();

            bool isSent = await _emailService.SendOtpAsync(newUser.Email, otpCode);

            if (!isSent)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResult("Tạo tài khoản thành công nhưng hệ thống gửi mail đang lỗi."));
            }

            _ = _emailService.SendWelcomeEmailAsync(newUser.Email, newUser.FullName);
            return Ok(ApiResponse<object>.SuccessResult(new { UserId = newUser.Id }, "Đăng ký thành công! Vui lòng kiểm tra email để nhận mã OTP."));
        }
        // ===========================
        // 1.5. VERIFY EMAIL (account created and saved to db immediately, if user cannot receive otp, go to API 1.7)
        // ===========================
        [HttpPost("verify-email")]
        public async Task<IActionResult> VerifyEmail([FromBody] OtpVerifyDto request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == request.Email);

            if (user == null)
            {
                return NotFound(ApiResponse<object>.ErrorResult("Không tìm thấy người dùng với email này."));
            }

            if (user.IsEmailVerified)
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Tài khoản này đã được xác thực trước đó."));
            }

            if (user.OtpCode != request.Otp || user.OtpExpiry < DateTime.UtcNow)
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Mã OTP không hợp lệ hoặc đã hết hạn."));
            }

            user.IsEmailVerified = true;
            user.OtpCode = default;
            user.OtpExpiry = default; 
            user.Status = UserStatus.ACTIVE;
            user.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return Ok(ApiResponse<object>.SuccessResult(null!, "Xác thực email thành công! Bạn đã có thể đăng nhập."));
        }
        // ===========================
        // 1.7. RESEND REGISTER OTP (When there is a user's email exist in db didn't receive the otp and need to resend during registration)
        // ONLY FOR UNVERIFIED EMAIL (isEmailVerified = false)
        // ===========================
        [HttpPost("resend-register-otp")]
        public async Task<IActionResult> ResendRegisterOtp([FromBody] ResendOtpDto request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == request.Email);

            if (user == null)
            {
                return NotFound(ApiResponse<object>.ErrorResult("Không tìm thấy tài khoản nào đăng ký với email này."));
            }

            if (user.IsEmailVerified)
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Tài khoản này đã được xác thực thành công. Vui lòng đăng nhập thẳng vào hệ thống!"));
            }

            if (user.OtpExpiry.HasValue && user.OtpExpiry.Value > DateTime.UtcNow.AddMinutes(4))
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Vui lòng đợi 60 giây trước khi yêu cầu gửi lại mã OTP mới."));
            }

            // 1. Generate new OTP code with 6 random digits
            string otpCode = new Random().Next(100000, 999999).ToString();

            // 2. Save OTP and expired after 5 minutes in database
            user.OtpCode = otpCode;
            user.OtpExpiry = DateTime.UtcNow.AddMinutes(5);
            user.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            // 3. Brevo email sending
            bool isSent = await _emailService.SendOtpAsync(user.Email, otpCode);

            if (!isSent)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResult("Tạo mã thành công nhưng hệ thống gửi mail đang gặp sự cố. Vui lòng thử lại sau."));
            }

            return Ok(ApiResponse<object>.SuccessResult(null!, "Đã gửi lại mã xác nhận OTP. Vui lòng kiểm tra cả hòm thư Spam (Thư rác)."));
        }
        // ===========================
        // 2. LOGIN
        // ===========================
        [HttpPost("login")]
        public async Task<IActionResult> Login([FromBody] LoginDto request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == request.Email);

            if (user == null || string.IsNullOrEmpty(user.PasswordHash))
            {
                return Unauthorized(ApiResponse<object>.ErrorResult("Email hoặc mật khẩu không đúng."));
            }

            bool isPasswordValid = false;
            try
            {
                if (user.PasswordHash.StartsWith("$2")) // BCrypt format
                {
                    isPasswordValid = BCrypt.Net.BCrypt.Verify(request.Password, user.PasswordHash);
                }
                else if (user.PasswordHash.Contains(".")) // PBKDF2 format from InternalUserService
                {
                    isPasswordValid = VerifyPbkdf2(request.Password, user.PasswordHash);
                }
            }
            catch
            {
                isPasswordValid = false;
            }

            if (!isPasswordValid)
            {
                return Unauthorized(ApiResponse<object>.ErrorResult("Email hoặc mật khẩu không đúng."));
            }

            // Check Verify Email
            if (!user.IsEmailVerified)
            {
                return BadRequest(new
                {
                    Success = false,
                    Message = "Tài khoản chưa được xác thực. Vui lòng kiểm tra email để lấy mã OTP xác nhận!",
                    Data = new { email = user.Email, isEmailVerified = false } // Trả về data để Front-end biết nhảy trang
                });
            }

            if (user.Status != UserStatus.ACTIVE)
            {
                return Unauthorized(ApiResponse<object>.ErrorResult("Tài khoản đã bị khóa."));
            }

            // Create token
            var token = _jwtService.GenerateToken(user);

            // Encapsulate response in LoginResponseDto format (Profile)
            var loginResponse = CreateLoginResponse(user, token);

            return Ok(ApiResponse<LoginResponseDto>.SuccessResult(loginResponse, "Đăng nhập thành công!"));
        }

        // ===========================
        // 3. GOOGLE LOGIN
        // ===========================
        [HttpPost("google-login")]
        public async Task<IActionResult> GoogleLogin([FromBody] GoogleLoginDto request)
        {
            try
            {
                var clientId = _configuration["GoogleAuth:ClientId"] ?? "TAM_THOI_CHUA_CO";

                var settings = new GoogleJsonWebSignature.ValidationSettings()
                {
                    Audience = new List<string>() { clientId }
                };

                // Verify Google Token
                var payload = await GoogleJsonWebSignature.ValidateAsync(request.IdToken, settings);

                // Find user by Gmail
                var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == payload.Email);

                if (user == null)
                {
                    // Auto-register if not exist
                    user = new UserModel
                    {
                        Email = payload.Email,
                        FullName = payload.Name,
                        Phone = null,
                        PasswordHash = "",
                        Role = UserRole.MEMBER,
                        Status = UserStatus.ACTIVE,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow,
                        IsEmailVerified = payload.EmailVerified
                    };

                    _context.Users.Add(user);
                    await _context.SaveChangesAsync();
                }
                else if (user.Status != UserStatus.ACTIVE)
                {
                    return Unauthorized(ApiResponse<object>.ErrorResult("Tài khoản của bạn đã bị khóa."));
                }

                // Local token generation
                var token = _jwtService.GenerateToken(user);

                // Encapsulate response in LoginResponseDto format
                var loginResponse = CreateLoginResponse(user, token);

                return Ok(ApiResponse<LoginResponseDto>.SuccessResult(loginResponse, "Đăng nhập Google thành công!"));
            }
            catch (InvalidJwtException)
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Token Google không hợp lệ hoặc đã hết hạn."));
            }
            catch (Exception ex)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResult($"Lỗi server: {ex.Message}"));
            }
        }

        // ===========================
        // Helper response encapsulation
        // ===========================
        private LoginResponseDto CreateLoginResponse(UserModel user, string token)
        {
            var userResponse = new UserResponseDto
            {
                Id = user.Id,
                Email = user.Email,
                FullName = user.FullName,
                Phone = user.Phone,
                Role = user.Role,
                Status = user.Status,
                CreatedAt = user.CreatedAt
            };

            return new LoginResponseDto
            {
                Token = token,
                User = userResponse,
                ExpiresAt = DateTime.UtcNow.AddDays(7) // Match the JWT expiration in JwtService
            };
        }

        // ===========================
        // UPDATE PROFILE (Name, Phone only)
        // ===========================
        [Authorize]
        [HttpPut("profile")]
        public async Task<IActionResult> UpdateProfile([FromBody] UpdateProfileDto request)
        {
            var userId = User.FindFirstValue(ClaimTypes.NameIdentifier);
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(ApiResponse<object>.ErrorResult("Không thể xác thực người dùng."));
            }

            var user = await _context.Users.FindAsync(userId);
            if (user == null)
            {
                return NotFound(ApiResponse<object>.ErrorResult("Không tìm thấy người dùng."));
            }

            user.FullName = request.FullName;
            user.Phone = request.Phone;

            _context.Users.Update(user);
            await _context.SaveChangesAsync();

            var updatedUserDto = new UserResponseDto
            {
                Id = user.Id,
                Email = user.Email,
                FullName = user.FullName,
                Phone = user.Phone,
                Role = user.Role,
                Status = user.Status,
                CreatedAt = user.CreatedAt
            };

            return Ok(ApiResponse<UserResponseDto>.SuccessResult(updatedUserDto, "Cập nhật thông tin thành công!"));
        }

        // ===========================
        // 4. FORGOT PASSWORD (OTP)
        // ===========================
        [HttpPost("forgot-password")]
        public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordDto request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == request.Email);

            if (user == null)
            {
                return Ok(ApiResponse<object>.SuccessResult(null!, "Nếu email tồn tại trong hệ thống, mã OTP đã được gửi."));
            }

            // If otpexpiry > 4 minutes, it means user just requested OTP and need to wait until 5 minutes to request again. This is to prevent spamming OTP request.
            if (user.OtpExpiry.HasValue && user.OtpExpiry.Value > DateTime.UtcNow.AddMinutes(4))
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Vui lòng đợi 60 giây trước khi yêu cầu gửi lại mã OTP mới."));
            }

            // 6 random digits OTP generation
            string otpCode = new Random().Next(100000, 999999).ToString();

            // Save OTP and expired after 5 minutes in database
            user.OtpCode = otpCode;
            user.OtpExpiry = DateTime.UtcNow.AddMinutes(5);
            await _context.SaveChangesAsync();

            bool isSent = await _emailService.SendOtpAsync(user.Email, otpCode);

            if (!isSent)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResult("Lỗi cấu hình gửi email. Vui lòng kiểm tra lại tài khoản SMTP."));
            }

            return Ok(ApiResponse<object>.SuccessResult(null!, "Đã gửi mã xác nhận đến email của bạn."));
        }
        // 5. RESET PASSWORD
        // ===========================
        [HttpPost("reset-password")]
        public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordDto request)
        {
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == request.Email);

            // Check if user exists and OTP is valid
            if (user == null || user.OtpCode != request.Otp || user.OtpExpiry < DateTime.UtcNow)
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Mã OTP không hợp lệ hoặc đã hết hạn."));
            }

            // Hashing new password and save to database
            user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(request.NewPassword);

            // Clear OTP fields after successful password reset
            user.OtpCode = default;
            user.OtpExpiry = default;
            user.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return Ok(ApiResponse<object>.SuccessResult(null!, "Đặt lại mật khẩu thành công! Bạn có thể đăng nhập bằng mật khẩu mới."));
        }
        // ===========================
        // 6. CHANGE PASSWORD
        // ===========================
        [Authorize] // Need token
        [HttpPost("change-password")]
        public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordDto request)
        {
            // Claim email from token
            var email = User.FindFirstValue(ClaimTypes.Email);
            if (string.IsNullOrEmpty(email))
            {
                return Unauthorized(ApiResponse<object>.ErrorResult("Không tìm thấy thông tin xác thực."));
            }

            var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == email);
            if (user == null)
            {
                return NotFound(ApiResponse<object>.ErrorResult("Không tìm thấy người dùng."));
            }

            // Block if user logged in with Google and has no hashpassword
            if (string.IsNullOrEmpty(user.PasswordHash))
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Tài khoản của bạn đăng nhập bằng Google nên chưa có mật khẩu. Vui lòng sử dụng tính năng 'Quên mật khẩu' để thiết lập mật khẩu mới."));
            }

            // Check old password
            bool isOldPasswordValid = false;
            try
            {
                if (user.PasswordHash.StartsWith("$2"))
                {
                    isOldPasswordValid = BCrypt.Net.BCrypt.Verify(request.OldPassword, user.PasswordHash);
                }
                else if (user.PasswordHash.Contains("."))
                {
                    isOldPasswordValid = VerifyPbkdf2(request.OldPassword, user.PasswordHash);
                }
            }
            catch { isOldPasswordValid = false; }

            if (!isOldPasswordValid)
            {
                return BadRequest(ApiResponse<object>.ErrorResult("Mật khẩu hiện tại không chính xác."));
            }

            // Save and hashing new password to database
            user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(request.NewPassword);
            user.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return Ok(ApiResponse<object>.SuccessResult(null!, "Đổi mật khẩu thành công!"));
        }

        private static bool VerifyPbkdf2(string password, string storedHash)
        {
            var parts = storedHash.Split('.');
            if (parts.Length != 3) return false;
            
            var iterations = int.Parse(parts[0]);
            var salt = Convert.FromBase64String(parts[1]);
            var hash = Convert.FromBase64String(parts[2]);
            
            using var pbkdf2 = new System.Security.Cryptography.Rfc2898DeriveBytes(password, salt, iterations, System.Security.Cryptography.HashAlgorithmName.SHA256);
            var computedHash = pbkdf2.GetBytes(32);
            
            return System.Security.Cryptography.CryptographicOperations.FixedTimeEquals(hash, computedHash);
        }
    }
}

File: back-end-nodejs/src/controllers/authController.js
import { Router } from 'express';
import bcrypt from 'bcryptjs';
import { User } from '../models/index.js';
import { UserRole, UserStatus } from '../models/enums.js';
import { ApiResponse } from '../dtos/index.js';
import { authenticate } from '../middlewares/index.js';

/**
 * Auth Controller - Tương đương AuthController.cs
 * Routes: /api/auth
 */
export function createAuthRouter(jwtService, emailService) {
  const router = Router();

  // ===========================
  // 1. REGISTER
  // ===========================
  router.post('/register', async (req, res) => {
    try {
      const { Email, Password, FullName, Phone } = req.body;

      // Check email exists
      const existingUser = await User.findOne({ email: Email?.toLowerCase() });
      if (existingUser) {
        return res.status(400).json(ApiResponse.error('Email này đã được sử dụng.'));
      }

      const otpCode = Math.floor(100000 + Math.random() * 900000).toString();
      const passwordHash = await bcrypt.hash(Password, 10);

      const newUser = await User.create({
        email: Email?.toLowerCase(),
        fullName: FullName,
        phone: Phone || '',
        passwordHash,
        role: UserRole.MEMBER,
        status: UserStatus.ACTIVE,
        isEmailVerified: false,
        otpCode,
        otpExpiry: new Date(Date.now() + 5 * 60 * 1000), // 5 minutes
      });

      // Send OTP email
      let isSent = true;
      try {
        isSent = await emailService.sendOtp(newUser.email, otpCode);
      } catch (err) {
        console.warn('Failed to send OTP email:', err.message);
        isSent = false;
      }

      if (!isSent) {
        return res.status(500).json(
          ApiResponse.error('Tạo tài khoản thành công nhưng hệ thống gửi mail đang lỗi.')
        );
      }

      // Send welcome email (non-blocking)
      emailService.sendWelcomeEmail(newUser.email, newUser.fullName).catch(() => { });

      return res.status(200).json(
        ApiResponse.success(
          { UserId: newUser._id.toString() },
          'Đăng ký thành công! Vui lòng kiểm tra email để nhận mã OTP.'
        )
      );
    } catch (error) {
      console.error('Register error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // 1.5. VERIFY EMAIL
  // ===========================
  router.post('/verify-email', async (req, res) => {
    try {
      const { Email, Otp } = req.body;

      const user = await User.findOne({ email: Email?.toLowerCase() });
      if (!user) {
        return res.status(404).json(ApiResponse.error('Không tìm thấy người dùng với email này.'));
      }

      if (user.isEmailVerified) {
        return res.status(400).json(ApiResponse.error('Tài khoản này đã được xác thực trước đó.'));
      }

      if (user.otpCode !== Otp || !user.otpExpiry || user.otpExpiry < new Date()) {
        return res.status(400).json(ApiResponse.error('Mã OTP không hợp lệ hoặc đã hết hạn.'));
      }

      user.isEmailVerified = true;
      user.otpCode = null;
      user.otpExpiry = null;
      user.status = UserStatus.ACTIVE;
      await user.save();

      return res.status(200).json(
        ApiResponse.success(null, 'Xác thực email thành công! Bạn đã có thể đăng nhập.')
      );
    } catch (error) {
      console.error('VerifyEmail error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // 1.7. RESEND REGISTER OTP
  // ===========================
  router.post('/resend-register-otp', async (req, res) => {
    try {
      const { Email } = req.body;

      const user = await User.findOne({ email: Email?.toLowerCase() });
      if (!user) {
        return res.status(404).json(
          ApiResponse.error('Không tìm thấy tài khoản nào đăng ký với email này.')
        );
      }

      if (user.isEmailVerified) {
        return res.status(400).json(
          ApiResponse.error('Tài khoản này đã được xác thực thành công. Vui lòng đăng nhập thẳng vào hệ thống!')
        );
      }

      // Throttle: 60 second cooldown
      if (user.otpExpiry && user.otpExpiry > new Date(Date.now() + 4 * 60 * 1000)) {
        return res.status(400).json(
          ApiResponse.error('Vui lòng đợi 60 giây trước khi yêu cầu gửi lại mã OTP mới.')
        );
      }

      const otpCode = Math.floor(100000 + Math.random() * 900000).toString();
      user.otpCode = otpCode;
      user.otpExpiry = new Date(Date.now() + 5 * 60 * 1000);
      await user.save();

      const isSent = await emailService.sendOtp(user.email, otpCode);
      if (!isSent) {
        return res.status(500).json(
          ApiResponse.error('Tạo mã thành công nhưng hệ thống gửi mail đang gặp sự cố. Vui lòng thử lại sau.')
        );
      }

      return res.status(200).json(
        ApiResponse.success(null, 'Đã gửi lại mã xác nhận OTP. Vui lòng kiểm tra cả hòm thư Spam (Thư rác).')
      );
    } catch (error) {
      console.error('ResendRegisterOtp error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // 2. LOGIN
  // ===========================
  router.post('/login', async (req, res) => {
    try {
      const { Email, Password } = req.body;

      const user = await User.findOne({ email: Email?.toLowerCase() });
      if (!user || !user.passwordHash || !(await bcrypt.compare(Password, user.passwordHash))) {
        return res.status(401).json(ApiResponse.error('Email hoặc mật khẩu không đúng.'));
      }

      // Check Verify Email
      if (!user.isEmailVerified) {
        return res.status(400).json({
          Success: false,
          Message: 'Tài khoản chưa được xác thực. Vui lòng kiểm tra email để lấy mã OTP xác nhận!',
          Data: { email: user.email, isEmailVerified: false },
        });
      }

      if (user.status !== UserStatus.ACTIVE) {
        return res.status(401).json(ApiResponse.error('Tài khoản đã bị khóa.'));
      }

      const token = jwtService.generateToken(user);
      const loginResponse = createLoginResponse(user, token);

      return res.status(200).json(ApiResponse.success(loginResponse, 'Đăng nhập thành công!'));
    } catch (error) {
      console.error('Login error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // 3. GOOGLE LOGIN
  // ===========================
  router.post('/google-login', async (req, res) => {
    try {
      const { IdToken } = req.body;

      // Dynamic import google-auth-library
      let OAuth2Client;
      try {
        const gAuth = await import('google-auth-library');
        OAuth2Client = gAuth.OAuth2Client;
      } catch {
        return res.status(500).json(ApiResponse.error('Google Auth library not available'));
      }

      const clientId = process.env.GOOGLE_CLIENT_ID || 'TAM_THOI_CHUA_CO';
      const client = new OAuth2Client(clientId);

      const ticket = await client.verifyIdToken({
        idToken: IdToken,
        audience: clientId,
      });
      const payload = ticket.getPayload();

      let user = await User.findOne({ email: payload.email });

      if (!user) {
        // Auto-register
        user = await User.create({
          email: payload.email,
          fullName: payload.name || '',
          phone: null,
          passwordHash: '',
          role: UserRole.MEMBER,
          status: UserStatus.ACTIVE,
          isEmailVerified: payload.email_verified || false,
        });
      } else if (user.status !== UserStatus.ACTIVE) {
        return res.status(401).json(ApiResponse.error('Tài khoản của bạn đã bị khóa.'));
      }

      const token = jwtService.generateToken(user);
      const loginResponse = createLoginResponse(user, token);

      return res.status(200).json(ApiResponse.success(loginResponse, 'Đăng nhập Google thành công!'));
    } catch (error) {
      console.error('Google login error:', error);
      if (error.message?.includes('Token') || error.message?.includes('token')) {
        return res.status(400).json(ApiResponse.error('Token Google không hợp lệ hoặc đã hết hạn.'));
      }
      return res.status(500).json(ApiResponse.error(`Lỗi server: ${error.message}`));
    }
  });

  // ===========================
  // 4. FORGOT PASSWORD (OTP)
  // ===========================
  router.post('/forgot-password', async (req, res) => {
    try {
      const { Email } = req.body;

      const user = await User.findOne({ email: Email?.toLowerCase() });
      if (!user) {
        // Security: don't reveal if email exists
        return res.status(200).json(
          ApiResponse.success(null, 'Nếu email tồn tại trong hệ thống, mã OTP đã được gửi.')
        );
      }

      // Throttle: 60 second cooldown
      if (user.otpExpiry && user.otpExpiry > new Date(Date.now() + 4 * 60 * 1000)) {
        return res.status(400).json(
          ApiResponse.error('Vui lòng đợi 60 giây trước khi yêu cầu gửi lại mã OTP mới.')
        );
      }

      const otpCode = Math.floor(100000 + Math.random() * 900000).toString();
      user.otpCode = otpCode;
      user.otpExpiry = new Date(Date.now() + 5 * 60 * 1000);
      await user.save();

      const isSent = await emailService.sendOtp(user.email, otpCode);
      if (!isSent) {
        return res.status(500).json(
          ApiResponse.error('Lỗi cấu hình gửi email. Vui lòng kiểm tra lại tài khoản SMTP.')
        );
      }

      return res.status(200).json(
        ApiResponse.success(null, 'Đã gửi mã xác nhận đến email của bạn.')
      );
    } catch (error) {
      console.error('ForgotPassword error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // 5. RESET PASSWORD
  // ===========================
  router.post('/reset-password', async (req, res) => {
    try {
      const { Email, Otp, NewPassword } = req.body;

      const user = await User.findOne({ email: Email?.toLowerCase() });
      if (!user || user.otpCode !== Otp || !user.otpExpiry || user.otpExpiry < new Date()) {
        return res.status(400).json(ApiResponse.error('Mã OTP không hợp lệ hoặc đã hết hạn.'));
      }

      user.passwordHash = await bcrypt.hash(NewPassword, 10);
      user.otpCode = null;
      user.otpExpiry = null;
      await user.save();

      return res.status(200).json(
        ApiResponse.success(null, 'Đặt lại mật khẩu thành công! Bạn có thể đăng nhập bằng mật khẩu mới.')
      );
    } catch (error) {
      console.error('ResetPassword error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // 6. CHANGE PASSWORD
  // ===========================
  router.post('/change-password', authenticate, async (req, res) => {
    try {
      const email = req.user?.email;
      if (!email) {
        return res.status(401).json(ApiResponse.error('Không tìm thấy thông tin xác thực.'));
      }

      const user = await User.findOne({ email: email.toLowerCase() });
      if (!user) {
        return res.status(404).json(ApiResponse.error('Không tìm thấy người dùng.'));
      }

      // Block if user logged in with Google and has no password
      if (!user.passwordHash) {
        return res.status(400).json(
          ApiResponse.error(
            "Tài khoản của bạn đăng nhập bằng Google nên chưa có mật khẩu. Vui lòng sử dụng tính năng 'Quên mật khẩu' để thiết lập mật khẩu mới."
          )
        );
      }

      const { OldPassword, NewPassword } = req.body;

      const isValid = await bcrypt.compare(OldPassword, user.passwordHash);
      if (!isValid) {
        return res.status(400).json(ApiResponse.error('Mật khẩu hiện tại không chính xác.'));
      }

      user.passwordHash = await bcrypt.hash(NewPassword, 10);
      await user.save();

      return res.status(200).json(ApiResponse.success(null, 'Đổi mật khẩu thành công!'));
    } catch (error) {
      console.error('ChangePassword error:', error);
      return res.status(500).json(ApiResponse.error('Internal server error'));
    }
  });

  // ===========================
  // Helper
  // ===========================
  function createLoginResponse(user, token) {
    return {
      Token: token,
      User: {
        Id: user._id.toString(),
        Email: user.email,
        FullName: user.fullName,
        Phone: user.phone,
        Role: user.role,
        Status: user.status,
        CreatedAt: user.createdAt,
      },
      ExpiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000), // 7 days
    };
  }

  return router;
}
